src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -176,6 +203,60 @@ public void mutuallyAuthenticate(APDU apdu) {
     respond(apdu, len, ISO7816.SW_NO_ERROR);
   }
 
+  /**
+   * Processes the LOAD_CERTS command. Copies the APDU buffer into `certs`.
+   * This function expects a DER signature and may only be called once.
+   * @param apdu the JCRE-owned APDU object.
+   */
+  public void loadCert(APDU apdu) {
+    byte[] apduBuffer = apdu.getBuffer();
+
+    if (idCertStatus != ID_CERTIFICATE_EMPTY) {
+      // Card cert may only be set once and never overwritten
+      ISOException.throwIt(ISO7816.SW_COMMAND_NOT_ALLOWED);
+    }
+
+    // Save the certificate
+    if (apduBuffer[ISO7816.OFFSET_LC] <= (byte)CERTIFICATE_MAX_LEN) {
+      Util.arrayCopyNonAtomic(apduBuffer, ISO7816.OFFSET_CDATA, idCertificate, (short) 0, (short) (apduBuffer[ISO7816.OFFSET_LC] & 0xff));
+      idCertStatus = ID_CERTIFICATE_LOCKED;
+    } else {
+      ISOException.throwIt(ISO7816.SW_DATA_INVALID);
+    }
+  }
+  
+  /**
+   * Processes the IDENTIFY_CARD command. Returns the card public key, and a signature on the
+   * challenge salt, to prove ownership of the key.
+   * @param apdu the JCRE-owned APDU object.
+   */
+  public void identifyCard(APDU apdu) {
+    byte[] apduBuffer = apdu.getBuffer();
+
+    // Ensure the received challenge is appropriate length
+    if (apduBuffer[ISO7816.OFFSET_LC] != (byte) MessageDigest.LENGTH_SHA_256) {
+      ISOException.throwIt(ISO7816.SW_DATA_INVALID);
+    }
+
+    short responseStart = (short) ISO7816.OFFSET_CDATA + (short) MessageDigest.LENGTH_SHA_256;
+    short off = responseStart;
+
+    // Copy card ID pubKey to the response buffer
+    ECPublicKey pk = (ECPublicKey) idKeypair.getPublic();
+    short pubkeyLen = pk.getW(apduBuffer, (short) (off + 2)); // Copy pubkey after TLV type and len
+    apduBuffer[off++] = (byte) 0x80; // TLV pubkey type
```

Comment:
Because I don't know `pubkeyLen` until I call `getW`.

I get that I could have just hard-coded `len` to 65 bytes, but my hope was to keep things fully constrained to programmable outputs where possible. This way, if somehow we end up with a 33 byte compressed key, we wont accidentally copy out 31 extra bytes of "unknown data"

Reply:
Okay, no problem